File: backend/consensus.py
"""Consensus secondary structure reconciliation logic.

Combines TANGO (aggregation) and S4PRED (secondary structure) predictions
into a tiered evidence framework. Based on AMYLPRED2 consensus approach
(Hamodrakas 2007) — "conformational switch zones" where helix-predicted
residues overlap beta-aggregation hotspots.

This module is pure functions with no side effects or I/O.
"""

from __future__ import annotations

from collections import Counter
from typing import Optional


def dominant_ss_at_region(
    ss_prediction: Optional[list[str]],
    hotspot_start: int,
    hotspot_end: int,
) -> str:
    """Find majority secondary structure call in a TANGO hotspot region.

    Args:
        ss_prediction: Per-residue SS predictions ('H', 'E', 'C').
        hotspot_start: Start index (inclusive).
        hotspot_end: End index (exclusive).

    Returns:
        Dominant SS call ('H', 'E', or 'C'). Defaults to 'C' if no data.
    """
    if not ss_prediction:
        return "C"

    # Clamp to valid bounds
    start = max(0, hotspot_start)
    end = min(len(ss_prediction), hotspot_end)

    region = ss_prediction[start:end]
    if not region:
        return "C"

    counts = Counter(region)
    # Sort by count descending, then alphabetically for tie-breaking
    return sorted(counts.keys(), key=lambda k: (-counts[k], k))[0]


def get_consensus_ss(
    tango_agg_max: Optional[float],
    s4pred_ss_at_hotspot: Optional[str],
    s4pred_helix_percent: Optional[float],
    ssw_prediction: Optional[int],
    s4pred_ssw_prediction: Optional[int],
    agg_threshold: float = 5.0,
    sequence_length: Optional[int] = None,
) -> dict:
    """Compute consensus tier from TANGO + S4PRED data.

    Tier Logic (the "Meytal" Rules):
        1: TANGO APR >threshold + S4PRED=Helix at hotspot → High-Confidence Switch Zone
        2: TANGO APR >threshold + S4PRED=Coil at hotspot  → Disordered Aggregation-Prone
        3: TANGO APR >threshold + S4PRED=Beta at hotspot  → Native Beta / Low Switch Risk
        4: TANGO APR <=threshold + any S4PRED              → No Aggregation Concern
        5: No TANGO data                                   → Insufficient Data

    Args:
        tango_agg_max: Peak TANGO aggregation propensity (%).
        s4pred_ss_at_hotspot: Dominant SS at TANGO hotspot ('H', 'E', 'C', or None).
        s4pred_helix_percent: S4PRED helix percentage (0-100).
        ssw_prediction: TANGO-based SSW prediction (1/-1/None).
        s4pred_ssw_prediction: S4PRED-based SSW prediction (1/-1/None).
        agg_threshold: TANGO aggregation threshold (default 5.0%).
        sequence_length: Sequence length for short-sequence penalty (optional).

    Returns:
        dict with keys: tier (1-5), label (str), certainty (0.0-1.0), explanation (str).
    """
    # ── Tier 5: No TANGO data ──
    if tango_agg_max is None:
        return {
            "tier": 5,
            "label": "Insufficient Data",
            "certainty": 0.0,
            "explanation": (
                "TANGO aggregation data is unavailable. Enable TANGO + S4PRED "
                "for consensus analysis."
            ),
        }

    # ── Tier 4: Low aggregation ──
    if tango_agg_max <= agg_threshold:
        certainty = 0.8
        certainty = _apply_ssw_modifier(certainty, ssw_prediction, s4pred_ssw_prediction)
        certainty = _apply_length_cap(certainty, sequence_length)
        return {
            "tier": 4,
            "label": "No Aggregation Concern",
            "certainty": certainty,
            "explanation": (
                f"TANGO peak aggregation ({tango_agg_max:.1f}%) is at or below "
                f"the {agg_threshold:.1f}% threshold. No significant aggregation "
                "hotspot detected."
            ),
        }

    # ── Tiers 1-3: High aggregation, differentiated by SS at hotspot ──
    ss = s4pred_ss_at_hotspot if s4pred_ss_at_hotspot in ("H", "E", "C") else "C"

    if ss == "H":
        tier, label, base_certainty = 1, "High-Confidence Switch Zone", 0.9
        explanation = (
            f"TANGO detects an aggregation hotspot (peak {tango_agg_max:.1f}%) "
            "where S4PRED predicts helical structure. This helix-to-beta "
            "conformational switch zone is a hallmark of amyloid-forming regions "
            "(Hamodrakas 2007)."
        )
    elif ss == "E":
        tier, label, base_certainty = 3, "Native Beta / Low Switch Risk", 0.5
        explanation = (
            f"TANGO detects an aggregation hotspot (peak {tango_agg_max:.1f}%) "
            "where S4PRED predicts beta-strand structure. The region is already "
            "in a beta conformation, so conformational switching is less likely."
        )
    else:  # "C" or None
        tier, label, base_certainty = 2, "Disordered Aggregation-Prone", 0.7
        explanation = (
            f"TANGO detects an aggregation hotspot (peak {tango_agg_max:.1f}%) "
            "in a disordered (coil) region. The lack of stable secondary structure "
            "may facilitate aggregation through disorder-to-order transitions."
        )

    certainty = _apply_ssw_modifier(base_certainty, ssw_prediction, s4pred_ssw_prediction)
    certainty = _apply_length_cap(certainty, sequence_length)

    return {
        "tier": tier,
        "label": label,
        "certainty": certainty,
        "explanation": explanation,
    }


def _apply_ssw_modifier(
    certainty: float,
    ssw_prediction: Optional[int],
    s4pred_ssw_prediction: Optional[int],
) -> float:
    """Adjust certainty based on SSW predictor agreement/disagreement."""
    if ssw_prediction is not None and s4pred_ssw_prediction is not None:
        if ssw_prediction == s4pred_ssw_prediction:
            certainty += 0.1  # agreement bonus
        else:
            certainty -= 0.1  # disagreement penalty
    return max(0.0, min(1.0, certainty))


def _apply_length_cap(certainty: float, sequence_length: Optional[int]) -> float:
    """Cap certainty for short sequences where S4PRED is out-of-distribution."""
    if sequence_length is not None and sequence_length < 20:
        certainty = min(certainty, 0.5)
    return certainty

File: ui/src/lib/consensus.ts
/**
 * Consensus secondary structure reconciliation logic.
 *
 * TypeScript port of backend/consensus.py — combines TANGO (aggregation)
 * and S4PRED (secondary structure) predictions into a tiered evidence
 * framework. Based on AMYLPRED2 consensus approach (Hamodrakas 2007).
 *
 * Operates on existing Peptide type fields — no API schema change needed.
 */

import type { Peptide } from "@/types/peptide";

export type ConsensusTier = 1 | 2 | 3 | 4 | 5;

export type ConsensusResult = {
  tier: ConsensusTier;
  label: string;
  certainty: number; // 0-1
  explanation: string;
  color: string; // tailwind border-color class
};

/** Tier metadata lookup. */
const TIER_META: Record<
  ConsensusTier,
  { label: string; color: string; baseCertainty: number }
> = {
  1: { label: "High-Confidence Switch Zone", color: "border-red-500", baseCertainty: 0.9 },
  2: { label: "Disordered Aggregation-Prone", color: "border-amber-500", baseCertainty: 0.7 },
  3: { label: "Native Beta / Low Switch Propensity", color: "border-blue-500", baseCertainty: 0.5 },
  4: { label: "Low Aggregation Propensity", color: "border-green-500", baseCertainty: 0.8 },
  5: { label: "Insufficient Data", color: "border-muted", baseCertainty: 0.0 },
};

/**
 * Find majority SS call ('H', 'E', 'C') in a region of per-residue predictions.
 */
export function dominantSsAtRegion(
  ssPrediction: string[] | undefined | null,
  start: number,
  end: number,
): string {
  if (!ssPrediction || ssPrediction.length === 0) return "C";

  const s = Math.max(0, start);
  const e = Math.min(ssPrediction.length, end);
  const region = ssPrediction.slice(s, e);
  if (region.length === 0) return "C";

  const counts: Record<string, number> = {};
  for (const ss of region) {
    counts[ss] = (counts[ss] ?? 0) + 1;
  }

  // Sort by count desc, then alphabetical for tie-breaking
  return Object.keys(counts).sort(
    (a, b) => counts[b] - counts[a] || a.localeCompare(b),
  )[0];
}

/**
 * Find the hotspot region (contiguous residues with agg > threshold).
 * Returns [start, end) indices of the peak hotspot region.
 */
function findHotspotRegion(
  aggCurve: number[] | undefined,
  threshold: number,
): [number, number] | null {
  if (!aggCurve || aggCurve.length === 0) return null;

  let bestStart = -1;
  let bestEnd = -1;
  let bestSum = 0;

  let curStart = -1;
  let curSum = 0;

  for (let i = 0; i < aggCurve.length; i++) {
    if (aggCurve[i] > threshold) {
      if (curStart === -1) curStart = i;
      curSum += aggCurve[i];
    } else {
      if (curStart !== -1 && curSum > bestSum) {
        bestStart = curStart;
        bestEnd = i;
        bestSum = curSum;
      }
      curStart = -1;
      curSum = 0;
    }
  }
  // Handle case where hotspot extends to end
  if (curStart !== -1 && curSum > bestSum) {
    bestStart = curStart;
    bestEnd = aggCurve.length;
  }

  return bestStart >= 0 ? [bestStart, bestEnd] : null;
}

/**
 * Compute consensus tier from a Peptide's TANGO + S4PRED data.
 */
export function getConsensusSS(
  peptide: Peptide,
  aggThreshold: number = 5.0,
): ConsensusResult {
  const tangoAggMax = peptide.tangoAggMax;

  // ── Tier 5: No TANGO data ──
  if (tangoAggMax == null) {
    return {
      tier: 5,
      label: TIER_META[5].label,
      certainty: 0.0,
      color: TIER_META[5].color,
      explanation:
        "TANGO aggregation data is unavailable. Enable TANGO + S4PRED for consensus analysis.",
    };
  }

  // ── Tier 4: Low aggregation ──
  if (tangoAggMax <= aggThreshold) {
    let certainty = TIER_META[4].baseCertainty;
    certainty = applySswModifier(certainty, peptide.sswPrediction, peptide.s4predSswPrediction);
    certainty = applyLengthCap(certainty, peptide.length);
    return {
      ...TIER_META[4],
      tier: 4,
      certainty,
      explanation: `TANGO peak aggregation (${tangoAggMax.toFixed(1)}%) is at or below the ${aggThreshold.toFixed(1)}% threshold. No significant aggregation hotspot detected.`,
    };
  }

  // ── Tiers 1-3: High aggregation, differentiated by SS at hotspot ──
  // Find SS at the hotspot region
  const hotspot = findHotspotRegion(peptide.tango?.agg, aggThreshold);
  const ssPred = peptide.s4pred?.ssPrediction;

  let ss: string;
  if (hotspot && ssPred) {
    ss = dominantSsAtRegion(ssPred, hotspot[0], hotspot[1]);
  } else {
    ss = "C"; // no S4PRED data → treat as disordered
  }

  let tier: ConsensusTier;
  let explanation: string;

  if (ss === "H") {
    tier = 1;
    explanation = `TANGO detects an aggregation hotspot (peak ${tangoAggMax.toFixed(1)}%) where S4PRED predicts helical structure. This helix-to-beta conformational switch zone is a hallmark of amyloid-forming regions (Hamodrakas 2007).`;
  } else if (ss === "E") {
    tier = 3;
    explanation = `TANGO detects an aggregation hotspot (peak ${tangoAggMax.toFixed(1)}%) where S4PRED predicts beta-strand structure. The region is already in a beta conformation, so conformational switching is less likely.`;
  } else {
    tier = 2;
    explanation = `TANGO detects an aggregation hotspot (peak ${tangoAggMax.toFixed(1)}%) in a disordered (coil) region. The lack of stable secondary structure may facilitate aggregation through disorder-to-order transitions.`;
  }

  let certainty = TIER_META[tier].baseCertainty;
  certainty = applySswModifier(certainty, peptide.sswPrediction, peptide.s4predSswPrediction);
  certainty = applyLengthCap(certainty, peptide.length);

  return {
    tier,
    label: TIER_META[tier].label,
    certainty,
    explanation,
    color: TIER_META[tier].color,
  };
}

function applySswModifier(
  certainty: number,
  sswPrediction: number | null | undefined,
  s4predSswPrediction: number | null | undefined,
): number {
  if (sswPrediction != null && s4predSswPrediction != null) {
    if (sswPrediction === s4predSswPrediction) {
      certainty += 0.1;
    } else {
      certainty -= 0.1;
    }
  }
  return Math.max(0.0, Math.min(1.0, certainty));
}

function applyLengthCap(certainty: number, length: number | null | undefined): number {
  if (length != null && length < 20) {
    certainty = Math.min(certainty, 0.5);
  }
  return certainty;
}
